stdlib/cache/src/main/java/org/ballerinalang/stdlib/cache/nativeimpl/Cache.java

```diff
@@ -31,52 +36,90 @@
+        if (cacheMap.size() >= capacity) {
+            int evictionKeysCount = (int) Math.ceil(capacity * evictionFactor);
+            cacheMap.setCapacity((capacity - evictionKeysCount));
+            cacheMap.setCapacity(capacity);
+        }
+        cacheMap.put(key, value);
     }
 
-    public static MapValue<String, Object> externGet(ObjectValue cache, String key) {
-        ConcurrentHashMap<String, MapValue<String, Object>> map =
-                (ConcurrentHashMap<String, MapValue<String, Object>>) cache.getNativeData(CACHE_MAP);
-        return map.get(key);
+    @SuppressWarnings("unchecked")
+    public static Object externGet(ObjectValue cache, String key, Long currentTime) {
+        cacheMap = (ConcurrentLinkedHashMap<String, BMap<String, Object>>) cache.getNativeData(CACHE);
+        BMap<String, Object> value = cacheMap.get(key);
+        Long time = (Long) value.get(EXPIRE_TIME);
+        if (time != -1 && time <= currentTime) {
+            cacheMap.remove(key);
+            return null;
+        }
+        return value;
     }
 
+    @SuppressWarnings("unchecked")
     public static void externRemove(ObjectValue cache, String key) {
-        ConcurrentHashMap<String, MapValue<String, Object>> map =
-                (ConcurrentHashMap<String, MapValue<String, Object>>) cache.getNativeData(CACHE_MAP);
-        map.remove(key);
+        cacheMap = (ConcurrentLinkedHashMap<String, BMap<String, Object>>) cache.getNativeData(CACHE);
+        cacheMap.remove(key);
     }
 
+    @SuppressWarnings("unchecked")
     public static void externRemoveAll(ObjectValue cache) {
-        ConcurrentHashMap<String, MapValue<String, Object>> map =
-                (ConcurrentHashMap<String, MapValue<String, Object>>) cache.getNativeData(CACHE_MAP);
-        map.clear();
+        cacheMap = (ConcurrentLinkedHashMap<String, BMap<String, Object>>) cache.getNativeData(CACHE);
+        cacheMap.clear();
     }
 
+    @SuppressWarnings("unchecked")
     public static boolean externHasKey(ObjectValue cache, String key) {
-        ConcurrentHashMap<String, MapValue<String, Object>> map =
-                (ConcurrentHashMap<String, MapValue<String, Object>>) cache.getNativeData(CACHE_MAP);
-        return map.containsKey(key);
+        cacheMap = (ConcurrentLinkedHashMap<String, BMap<String, Object>>) cache.getNativeData(CACHE);
+        return cacheMap.containsKey(key);
     }
 
-    public static ArrayValueImpl externKeys(ObjectValue cache) {
-        ConcurrentHashMap<String, MapValue<String, Object>> map =
-                (ConcurrentHashMap<String, MapValue<String, Object>>) cache.getNativeData(CACHE_MAP);
-        return new ArrayValueImpl(map.keySet().toArray(new String[0]));
+    @SuppressWarnings("unchecked")
+    public static BArray externKeys(ObjectValue cache) {
+        cacheMap = (ConcurrentLinkedHashMap<String, BMap<String, Object>>) cache.getNativeData(CACHE);
+        String[] keySets = cacheMap.keySet().toArray(new String[0]);
```

Comment:
Can't we return this as `return new ArrayValueImpl(cacheMap.keySet().toArray(new String[0]));` ?

Reply:
When changing to this, getting the following error.
`{ballerina}TypeCastError message=incompatible types: 'string[]' cannot be cast to 'handle[]'`